File: Algorithms/543.diameter-of-binary-tree/diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
    m:=0
    max:=&m  //现有的最长直径
    if root==nil{
        return 0
    }
    depth(root,max)
    return *max
}

func depth(root *TreeNode,max *int)int{
    if root==nil{
        return 0
    }
    a:=depth(root.Left,max)
    b:=depth(root.Right,max)
    *max =int(math.Max(float64(a+b),float64(*max)))//此节点的直径和现有的最长直径比较
    return int(math.Max(float64(a),float64(b)))+1//返回深度
}

File: lintcode/1181.diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: a root of binary tree
     * @return: return a integer
     */
    public int diameterOfBinaryTree(TreeNode root) {
        return getDeepth(root.left) + getDeepth(root.right);
    }
    //递归求树的高度
    private int getDeepth(TreeNode root){
        if(root == null){
            return 0;
        }
        return Math.max(getDeepth(root.left),getDeepth(root.right)) + 1;
    }
}